File: 2023/day02/day02.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	body, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Printf("failed to read file: %v\n", err.Error())
		return
	}

	bagSet := map[string]int{
		"red":   12,
		"green": 13,
		"blue":  14,
	}

	lines := strings.Split(strings.Trim(string(body), "\n"), "\n")
	sum := 0
	for _, line := range lines {
		line = line[5:]
		colonSeparated := strings.Split(line, ": ")
		gameId, _ := strconv.Atoi(colonSeparated[0])
		ballSets := strings.Split(colonSeparated[1], "; ")
		isValid := true
		for _, ballSet := range ballSets {
			balls := strings.Split(ballSet, ", ")
			for _, ball := range balls {
				spaceSeparated := strings.Split(ball, " ")
				amount, _ := strconv.Atoi(spaceSeparated[0])
				if amount > bagSet[spaceSeparated[1]] {
					isValid = false
					break
				}
			}

			if !isValid {
				break
			}
		}

		if isValid {
			sum += gameId
		}
	}

	fmt.Println(sum)
}

File: 2023/day02/day02.py
from collections import defaultdict


def main():
    with open("input.txt") as f:
        lines = f.readlines()

    sum = 0
    for line in lines:
        line = line[5:-1]
        colon_separated = line.split(": ")
        ball_sets = colon_separated[1].split("; ")
        curr_max = defaultdict(int)
        for ball_set in ball_sets:
            balls = ball_set.split(", ")
            for ball in balls:
                amount_str, colour = ball.split(" ")
                amount = int(amount_str)
                if amount > curr_max[colour]:
                    curr_max[colour] = amount

        power = curr_max["red"] * curr_max["green"] * curr_max["blue"]
        print(f"curr_max: {curr_max}, power: {power}")
        sum += power

    print(sum)


if __name__ == "__main__":
    main()